### PR TITLE
Add generic end 2 end test to all block transforms

### PR DESCRIPTION
**Pull request by jorgefilipecosta**
## Description
This PR proposes a generic approach to test the block transforms we have.
Fist it loads all the fixtures of content we have and checks if the transforms that should appear on the block they represent correctly appear.
Then it checks in a loop if each transform contains the output that should is expected.
If in the future a new block is created a new fixture should create for it. If that block contains transforms, we should update our fixtures/block-transforms.json specifying which transforms are available and a new test case is automatically generated and a snapshot created. if the fixtures/block-transforms.json file is not updated the tests will fail so this makes sure the tests do not get outdated, each transform is tested.

### CODE REVIEW

File: test/integration/full-content/fixtures/core__gallery.html [hunk 1]
```diff
@@ -7,7 +7,7 @@
 	</li>
 	<li class="blocks-gallery-item">
 		<figure>
-			<img src="http://google.com/hi.png" alt="title" />
+			<img src="https://cldup.com/cXyG__fTLN.jpg" alt="title" />
```
Comment (iseulde) [inline test/integration/full-content/fixtures/core__gallery.html]: Could you explain this change? Thanks!

Comment (iseulde) [reply test/integration/full-content/fixtures/core__gallery.html]: Oh I see now, never mind. Would be cool if the image moved so it's in our control.

File: test/e2e/specs/__snapshots__/block-transforms.test.js.snap [hunk 2]
```diff
@@ -0,0 +1,297 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`test transforms individual transforms work as expected Should correctly transform block in fixture core__audio to File block 1`] = `
```
Comment (iseulde) [inline test/e2e/specs/__snapshots__/block-transforms.test.js.snap]: Is there any way to make this more descriptive and concise? Something like: "Paragraph block should transform to Heading block"?

File: test/e2e/specs/block-transforms.test.js [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+	beforeAll( async () => {
+		await newPost();
+		await page.click( '.editor-post-title .editor-post-title__block' );
+		for ( const fileBase of fileBasenames ) {
+			transformStructure[ fileBase ] = await getTransformStructureFromFile(
+				fixturesDir,
+				fileBase
+			);
+		}
+		expect( console ).toHaveErroredWith(
```
Comment (iseulde) [inline test/e2e/specs/block-transforms.test.js]: Why are we expecting this? Could this be commented?

File: test/e2e/specs/block-transforms.test.js [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+	const fixturesDir = path.join(
+		__dirname, '..', '..', 'integration', 'full-content', 'fixtures'
+	);
+
+	const fileBasenames = getFileBaseNames( fixturesDir );
+
+	const transformStructure = {};
+	beforeAll( async () => {
+		await newPost();
+		await page.click( '.editor-post-title .editor-post-title__block' );
```
Comment (iseulde) [inline test/e2e/specs/block-transforms.test.js]: Why is this needed? Just wondering.

File: test/e2e/specs/block-transforms.test.js [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+				fixturesDir,
+				fileBase
+			);
+		}
+		expect( console ).toHaveErroredWith(
+			'Failed to load resource: the server responded with a status of 404 (Not Found)'
+		);
+	} );
+
+	it( 'Should contain the expected transforms', async () => {
```
Comment (iseulde) [inline test/e2e/specs/block-transforms.test.js]: Generally I think we should not capitalise "should" because it will be placed in a sentence.

File: test/e2e/specs/block-transforms.test.js [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+				),
+				'availableTransforms'
+			)
+		).toEqual( expectedTransforms );
+	} );
+
+	describe( 'individual transforms work as expected', () => {
+		afterEach( async () => {
+			await setPostContent( '' );
+			await page.click( '.editor-post-title .editor-post-title__block' );
```
Comment (iseulde) [inline test/e2e/specs/block-transforms.test.js]: After reading `beforeAll`, it should like this should happen `beforeEach`?

File: test/integration/full-content/fixtures/core__gallery.html (hunk 1 above)
Comment (jorgefilipecosta) [reply test/integration/full-content/fixtures/core__gallery.html]: Yes, that's true but what reliable source can we consider under our control?
I used cldup here as we are also using it on the demo page so at least we only depend on this domain.

File: test/integration/full-content/fixtures/core__video.html [hunk 7]
```diff
@@ -1,3 +1,3 @@
 <!-- wp:core/video -->
-<figure class="wp-block-video"><video controls src="https://awesome-fake.video/file.mp4"></video></figure>
+<figure class="wp-block-video"><video controls src="https://www.sample-videos.com/video/mp4/720/big_buck_bunny_720p_1mb.mp4"></video></figure>
```
[the remark this answers is not in the document]
Comment (jorgefilipecosta) [reply test/integration/full-content/fixtures/core__video.html]: I will switch to a cldup url.

File: test/e2e/specs/fixtures/block-transforms.json [hunk 8]
```diff
@@ -0,0 +1,105 @@
+{
+	"core__archives":[],
+	"core__archives__showPostCounts":[],
+	"core__audio":[
+		"File"
```
Comment (iseulde) [inline test/e2e/specs/fixtures/block-transforms.json]: Any way to have this both be block names (plus attributes) or both be file names?

File: test/support/utils.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
+/**
```
Comment (iseulde) [inline test/support/utils.js]: Is there any of this that could be shared with `full-content.spec.js`?

File: test/e2e/specs/__snapshots__/block-transforms.test.js.snap (hunk 2 above)
Comment (jorgefilipecosta) [reply test/e2e/specs/__snapshots__/block-transforms.test.js.snap]: We can do that, but the thing is that we have many tests testing Paragraph to Heading block (one for each paragraph fixture), that was the reason I added the fixture to make it easier to debug in case the test fails. Maybe adding also the block name is a good trade-off.

Comment (iseulde) [reply test/e2e/specs/__snapshots__/block-transforms.test.js.snap]: Ok, we can always iterate on this.

File: test/integration/full-content/fixtures/core__video.html (hunk 7 above)
[the remark this answers is not in the document]
Comment (iseulde) [reply test/integration/full-content/fixtures/core__video.html]: I don't know about cldup either tbh. I think any files should be in this repo. Test images an files could be super tiny.

[the remark this answers is not in the document]
Comment (iseulde) [reply test/integration/full-content/fixtures/core__video.html]: I guess it can be cloud up for now as we're already doing that.

[the remark this answers is not in the document]
Comment (iseulde) [reply test/integration/full-content/fixtures/core__video.html]: Tiny files might speed up the tests as well though.

[the remark this answers is not in the document]
Comment (iseulde) [reply test/integration/full-content/fixtures/core__video.html]: Idea: https://github.com/mathiasbynens/small.

[the remark this answers is not in the document]
Comment (jorgefilipecosta) [reply test/integration/full-content/fixtures/core__video.html]: I like this idea, but I'm not certain how we can make sure that a local URL for a file is available. The other option would use a data URL that contains the files inline but for example, a video still contains many characters.

[the remark this answers is not in the document]
Comment (iseulde) [reply test/integration/full-content/fixtures/core__video.html]: Apparently it would be:

```
data:video/mp4;base64,AAAAHGZ0eXBpc29tAAACAGlzb21pc28ybXA0MQAAAAhmcmVlAAAAGm1kYXQAAAGzABAHAAABthBgUYI9t+8AAAMNbW9vdgAAAGxtdmhkAAAAAMXMvvrFzL76AAAD6AAAACoAAQAAAQAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAgAAABhpb2RzAAAAABCAgIAHAE/////+/wAAAiF0cmFrAAAAXHRraGQAAAAPxcy++sXMvvoAAAABAAAAAAAAACoAAAAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAAAAAAAAABAAAAAAAAAAAAAAAAAABAAAAAAAgAAAAIAAAAAAG9bWRpYQAAACBtZGhkAAAAAMXMvvrFzL76AAAAGAAAAAEVxwAAAAAALWhkbHIAAAAAAAAAAHZpZGUAAAAAAAAAAAAAAABWaWRlb0hhbmRsZXIAAAABaG1pbmYAAAAUdm1oZAAAAAEAAAAAAAAAAAAAACRkaW5mAAAAHGRyZWYAAAAAAAAAAQAAAAx1cmwgAAAAAQAAAShzdGJsAAAAxHN0c2QAAAAAAAAAAQAAALRtcDR2AAAAAAAAAAEAAAAAAAAAAAAAAAAAAAAAAAgACABIAAAASAAAAAAAAAABAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAGP//AAAAXmVzZHMAAAAAA4CAgE0AAQAEgICAPyARAAAAAAMNQAAAAAAFgICALQAAAbABAAABtYkTAAABAAAAASAAxI2IAMUARAEUQwAAAbJMYXZjNTMuMzUuMAaAgIABAgAAABhzdHRzAAAAAAAAAAEAAAABAAAAAQAAABxzdHNjAAAAAAAAAAEAAAABAAAAAQAAAAEAAAAUc3RzegAAAAAAAAASAAAAAQAAABRzdGNvAAAAAAAAAAEAAAAsAAAAYHVkdGEAAABYbWV0YQAAAAAAAAAhaGRscgAAAAAAAAAAbWRpcmFwcGwAAAAAAAAAAAAAAAAraWxzdAAAACOpdG9vAAAAG2RhdGEAAAABAAAAAExhdmY1My4yMS4x
```

[the remark this answers is not in the document]
Comment (iseulde) [reply test/integration/full-content/fixtures/core__video.html]: Not a small amount of characters, but it still might have a big impact on test speed.

**Comment by jorgefilipecosta** [comment]
Hi @iseulde, thank you a lot for your review, I think all the feedback was applied.

**Comment by jorgefilipecosta** [comment]
Closed in favor of https://github.com/WordPress/gutenberg/pull/12336.
